File: 프로그래머스/2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.cpp
#include <vector>
#include <unordered_set>

using namespace std;

int solution(vector<int> elements) {
    int n = elements.size();
    unordered_set<int> sums;

    vector<int> ext(elements.begin(), elements.end());
    ext.insert(ext.end(), elements.begin(), elements.end());

    for (int length = 1; length <= n; ++length) {
        for (int start = 0; start < n; ++start) {
            int sum = 0;
            for (int i = 0; i < length; ++i) {
                sum += ext[start + i];
            }
            sums.insert(sum);
        }
    }

    return sums.size();
}

File: 프로그래머스/2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.java
import java.util.*;

class Solution {
    public int solution(int[] elements) {
        int size = elements.length;
        int[] numbers = new int[size * 2 - 1];
        for (int i = 0; i < size * 2 - 1; i++) {
            numbers[i] = elements[i%size];
        }
        
        Set<Integer> s = new HashSet<>();
        
        for (int l = 1; l <= size; l++) {
            int sum = 0;
            
            for (int i = 0; i < l; i++) {
                sum += numbers[i];
            }
            
            s.add(sum);
            
            for (int i = 1; i < size; i++) {
                sum = sum - numbers[i - 1] + numbers[i + l - 1];
                s.add(sum);
            }
        }
        return s.size();
    }
}